File: 2021/LevelUp_June/l002_Tree/l004Construction.java
import java.util.ArrayList;
import java.util.LinkedList;

public class l004Construction {
    public static class TreeNode {
        int val = 0;
        TreeNode left = null;
        TreeNode right = null;

        TreeNode(int val) {
            this.val = val;
        }
    }

    public static TreeNode constructFromInOrder(int[] inOrder, int si, int ei) {
        if (si > ei)
            return null;
        int mid = (si + ei) / 2;
        TreeNode root = new TreeNode(inOrder[mid]);

        root.left = constructFromInOrder(inOrder, si, mid - 1);
        root.right = constructFromInOrder(inOrder, mid + 1, ei);

        return root;
    }

    public static TreeNode constructFromInOrder(int[] inOrder) {
        return constructFromInOrder(inOrder, 0, inOrder.length - 1);
    }

    public class Codec {
        public void serialize(TreeNode root, StringBuilder sb) {
            if (root == null)
                return;

            sb.append(root.val + " ");
            serialize(root.left, sb);
            serialize(root.right, sb);
        }

        public String serialize(TreeNode root) {
            StringBuilder sb = new StringBuilder();
            serialize(root, sb);

            return sb.toString();
        }

        public TreeNode bstFromPreorder(int[] preorder, int lr, int rr, int[] idx) {
            int i = idx[0];
            if (i >= preorder.length || preorder[i] < lr || preorder[i] > rr)
                return null;

            TreeNode root = new TreeNode(preorder[i]);
            idx[0]++;

            root.left = bstFromPreorder(preorder, lr, root.val, idx);
            root.right = bstFromPreorder(preorder, root.val, rr, idx);

            return root;
        }

        // Decodes your encoded data to tree.
        public TreeNode deserialize(String data) {
            if (data.equals(""))
                return null;

            String[] arr = data.split(" ");
            int[] preorder = new int[arr.length];
            for (int i = 0; i < arr.length; i++) {
                preorder[i] = Integer.parseInt(arr[i]);
            }

            int[] idx = new int[1];
            return bstFromPreorder(preorder, -(int) 1e9, (int) 1e9, idx);
        }
    }

    public static TreeNode buildTree(int[] preorder, int psi, int pei, int[] inorder, int isi, int iei) {
        if (psi > pei)
            return null;

        int idx = isi;
        while (inorder[idx] != preorder[psi])
            idx++;

        int tnel = idx - isi;
        TreeNode root = new TreeNode(preorder[psi]);
        root.left = buildTree(preorder, psi + 1, psi + tnel, inorder, isi, idx - 1);
        root.right = buildTree(preorder, psi + tnel + 1, pei, inorder, idx + 1, iei);

        return root;
    }

    public static TreeNode buildTree(int[] preorder, int[] inorder) {
        int n = preorder.length;
        return buildTree(preorder, 0, n - 1, inorder, 0, n - 1);
    }

    public static TreeNode buildTreeFromPostAndIn(int[] postOrder, int psi, int pei, int[] inorder, int isi, int iei) {
        if (psi > pei)
            return null;

        int idx = isi;
        while (inorder[idx] != postOrder[pei])
            idx++;

        int tnel = idx - isi;
        TreeNode root = new TreeNode(postOrder[pei]);
        root.left = buildTreeFromPostAndIn(postOrder, psi, psi + tnel - 1, inorder, isi, idx - 1);
        root.right = buildTreeFromPostAndIn(postOrder, psi + tnel, pei - 1, inorder, idx + 1, iei);

        return root;
    }

    public static TreeNode buildTreeFromPostAndIn(int[] inorder, int[] postorder) {
        int n = inorder.length;
        return buildTreeFromPostAndIn(postorder, 0, n - 1, inorder, 0, n - 1);
    }

    public static TreeNode constructFromPrePost(int[] postOrder, int ppsi, int ppei, int[] preOrder, int psi, int pei) {
        if (psi > pei)
            return null;

        TreeNode root = new TreeNode(preOrder[psi]);
        if (psi == pei)
            return root;

        int idx = ppsi;
        while (postOrder[idx] != preOrder[psi + 1])
            idx++;

        int tnel = idx - ppsi + 1;
        root.left = constructFromPrePost(postOrder, ppsi, ppsi + tnel - 1, preOrder, psi + 1, psi + tnel);
        root.right = constructFromPrePost(postOrder, ppsi + tnel, pei - 1, preOrder, psi + tnel + 1, pei);

        return root;
    }

    public static TreeNode constructFromPrePost(int[] preOrder, int[] postorder) {
        int n = postorder.length;
        return constructFromPrePost(postorder, 0, n - 1, preOrder, 0, n - 1);
    }

    public class CodecBinaryTree {

        public void serialize(TreeNode root, StringBuilder sb) {
            if (root == null) {
                sb.append("# ");
                return;
            }

            sb.append(root.val + " ");
            serialize(root.left, sb);
            serialize(root.right, sb);
        }

        public String serialize(TreeNode root) {
            if (root == null)
                return "";
            StringBuilder sb = new StringBuilder();
            serialize(root, sb);

            return sb.toString();
        }

        int idx = 0;

        public TreeNode deserialize(String[] arr) {
            if (idx >= arr.length || arr[idx].equals("#")) {
                idx++;
                return null;
            }

            TreeNode root = new TreeNode(Integer.parseInt(arr[idx++]));
            root.left = deserialize(arr);
            root.right = deserialize(arr);

            return root;
        }

        public TreeNode deserialize(String data) {
            if (data.length() == 0)
                return null;

            idx = 0;
            String[] arr = data.split(" ");
            return deserialize(arr);
        }
    }

    public class CodecBinaryTree_02 {

        public String serialize(TreeNode root) {
            if (root == null)
                return "";
            StringBuilder sb = new StringBuilder();
            LinkedList<TreeNode> que = new LinkedList<>();
            que.addLast(root);

            while (que.size() != 0) {
                TreeNode rnode = que.removeFirst();
                sb.append((rnode != null ? rnode.val : "#") + " ");

                if (rnode == null)
                    continue;

                que.addLast(rnode.left);
                que.addLast(rnode.right);
            }

            return sb.toString();
        }

        public TreeNode deserialize(String data) {
            if (data.length() == 0)
                return null;

            String[] arr = data.split(" ");
            LinkedList<TreeNode> que = new LinkedList<>();
            TreeNode root = new TreeNode(Integer.parseInt(arr[0]));
            que.addLast(root);

            int idx = 1;
            while (que.size() != 0) {
                TreeNode rnode = que.removeFirst();

                if (!arr[idx].equals("#")) {
                    TreeNode leftChild = new TreeNode(Integer.parseInt(arr[idx]));
                    rnode.left = leftChild;
                    que.addLast(leftChild);
                }
                idx++;

                if (!arr[idx].equals("#")) {
                    TreeNode rightChild = new TreeNode(Integer.parseInt(arr[idx]));
                    rnode.right = rightChild;
                    que.addLast(rightChild);
                }
                idx++;

            }

            return root;
        }
    }

}

File: 2021/LevelUp_June/l002_Tree/l004construction.cpp
#include <iostream>
#include <vector>
#include <list>
#include <stringstream>

using namespace std;

class TreeNode
{
public:
    int val = 0;
    TreeNode *left = nullptr;
    TreeNode *right = nullptr;

    TreeNode(int val)
    {
        this->val = val;
    }
};

TreeNode *getMidNode(TreeNode *head)
{
    if (head == nullptr || head->right == nullptr)
        return head;
    TreeNode *slow = head, *fast = head;
    while (fast->right != nullptr && fast->right->right != nullptr)
    {
        fast = fast->right->right;
        slow = slow->right;
    }

    return slow;
}

TreeNode *dllToBST(TreeNode *head)
{
    if (head == nullptr || head->right == nullptr)
        return head;

    TreeNode *midNode = getMidNode(head);
    TreeNode *prev = midNode->left, *forw = midNode->right;

    midNode->left = midNode->right = forw->left = nullptr;
    if (prev != nullptr)
        prev->right = nullptr;

    TreeNode *root = midNode, *leftHead = (prev != nullptr ? head : nullptr), *rightHead = forw;

    root->left = dllToBST(leftHead);
    root->right = dllToBST(rightHead);

    return root;
}

TreeNode *mergeTwoSortedDLL(TreeNode *l1, TreeNode *l2)
{
    if (l1 == nullptr || l2 == nullptr)
        return l1 != nullptr ? l1 : l2;

    TreeNode *dummy = new TreeNode(-1);
    TreeNode *prev = dummy, *c1 = l1, *c2 = l2;

    while (c1 != nullptr && c2 != nullptr)
    {
        if (c1->val <= c2->val)
        {
            prev->right = c1;
            c1->left = prev;

            c1 = c1->right;
        }
        else
        {
            prev->right = c2;
            c2->left = prev;

            c2 = c2->right;
        }

        prev = prev->right;
    }

    if (c1 != nullptr)
    {
        prev->right = c1;
        c1->left = prev;
    }
    else
    {
        prev->right = c2;
        c2->left = prev;
    }

    TreeNode *head = dummy->right;
    dummy->right = head->left = nullptr;

    return head;
}

TreeNode *mergeSort(TreeNode *head)
{
    if (head == nullptr || head->right == nullptr)
        return head;

    TreeNode *midNode = getMidNode(head);
    TreeNode *forwHead = midNode->right;
    forwHead->left = midNode->right = nullptr;

    return mergeTwoSortedDLL(mergeSort(head), mergeSort(forwHead));
}

TreeNode *getRightMostNode(TreeNode *node, TreeNode *curr)
{
    while (node->right != nullptr && node->right != curr)
    {
        node = node->right;
    }

    return node;
}

TreeNode *dll(TreeNode *root)
{
    TreeNode *dummy = new TreeNode(-1);
    TreeNode *curr = root, *prev = dummy;
    while (curr != nullptr)
    {
        TreeNode *left = curr->left;
        if (left == nullptr)
        {
            prev->right = curr;
            curr->left = prev;
            prev = prev->right;

            curr = curr->right;
        }
        else
        {
            TreeNode *rightMostNode = getRightMostNode(left, curr);
            if (rightMostNode->right == nullptr)
            {
                rightMostNode->right = curr;
                curr = curr->left;
            }
            else
            {

                rightMostNode->right = nullptr;

                prev->right = curr;
                curr->left = prev;
                prev = prev->right;

                curr = curr->right;
            }
        }
    }

    TreeNode *head = dummy->right;
    dummy->right = head->left = nullptr;

    return head;
}

void display(TreeNode *node)
{
    if (node == nullptr)
        return;

    string str = "";
    str += ((node->left != nullptr ? to_string(node->left->val) : "."));
    str += (" -> " + to_string(node->val) + " <- ");
    str += ((node->right != nullptr ? to_string(node->right->val) : "."));

    cout << str << endl;

    display(node->left);
    display(node->right);
}

TreeNode *BTToBST(TreeNode *root)
{
    if (root == nullptr)
        return nullptr;

    TreeNode *head = dll(root); // Binary Tree to Doubly Lihnked List
    head = mergeSort(head);     // doubly Linked List to Sorted Doubly Linked List
    root = dllToBST(head);

    return root;
}

TreeNode *constructFromInOrder(vector<int> &inOrder, int si, int ei)
{
    if (si > ei)
        return nullptr;
    int mid = (si + ei) / 2;
    TreeNode *root = new TreeNode(inOrder[mid]);

    root->left = constructFromInOrder(inOrder, si, mid - 1);
    root->right = constructFromInOrder(inOrder, mid + 1, ei);

    return root;
}

TreeNode *constructFromInOrder(vector<int> &inOrder)
{
    return constructFromInOrder(inOrder, 0, inOrder.size() - 1);
}

// idx: {0 :: actual idx}
TreeNode *bstFromPreorder(vector<int> &preorder, int lr, int rr, vector<int> &idx)
{
    int i = idx[0];
    if (i >= preorder.size() || preorder[i] < lr || preorder[i] > rr)
        return nullptr;

    TreeNode *root = new TreeNode(preorder[i]);
    idx[0]++;

    root->left = bstFromPreorder(preorder, lr, root->val, idx);
    root->right = bstFromPreorder(preorder, root->val, rr, idx);

    return root;
}

TreeNode *bstFromPreorder(vector<int> &preorder)
{
    vector<int> idx(1, 0);
    return bstFromPreorder(preorder, -1e9, 1e9, idx);
}

// idx: {0 :: actual idx}
TreeNode *bstFromPostorder(vector<int> &postorder, int lr, int rr, vector<int> &idx)
{
    int i = idx[0];
    if (i <= -1 || postorder[i] < lr || postorder[i] > rr)
        return nullptr;

    TreeNode *root = new TreeNode(postorder[i]);
    idx[0]--;

    root->right = bstFromPostorder(postorder, root->val, rr, idx);
    root->left = bstFromPostorder(postorder, lr, root->val, idx);

    return root;
}

TreeNode *bstFromPreorder(vector<int> &postorder)
{
    vector<int> idx(1, postorder.size() - 1);
    return bstFromPreorder(postorder, -1e9, 1e9, idx);
}

class Codec
{
public:
    void serialize(TreeNode *root, string &sb)
    {
        if (root == nullptr)
            return;

        sb += to_string(root->val) + " ";
        serialize(root->left, sb);
        serialize(root->right, sb);
    }

    string serialize(TreeNode *root)
    {
        string sb = "";
        serialize(root, sb);
        return sb;
    }

    TreeNode *bstFromPreorder(vector<int> &preorder, int lr, int rr, vector<int> &idx)
    {
        int i = idx[0];
        if (i >= preorder.size() || preorder[i] < lr || preorder[i] > rr)
            return nullptr;

        TreeNode *root = new TreeNode(preorder[i]);
        idx[0]++;

        root->left = bstFromPreorder(preorder, lr, root->val, idx);
        root->right = bstFromPreorder(preorder, root->val, rr, idx);

        return root;
    }

    TreeNode *deserialize(string data)
    {
        if (data.length() == 0)
            return nullptr;
        stringstream ss(data);
        string word;

        vector<int> preorder;
        while (ss >> word)
        {
            int val = stoi(word);
            preorder.push_back(val);
        }

        vector<int> idx(1, 0);
        return bstFromPreorder(preorder, -(int)1e9, (int)1e9, idx);
    }
};

// for you :: https://www.pepcoding.com/resources/data-structures-and-algorithms-in-java-levelup/trees/construct-bst-from-levelorder-traversal/ojquestion
//for you :: https://www.lintcode.com/problem/1307/

class BSTPair
{
public:
    bool isBal = true;
    int h = -1;

    bool isBST = true;
    int min = (int)1e9;
    int max = -(int)1e9;

    int size = 0;
    TreeNode *largestRoot = nullptr;
};

BSTPair isBalanced_(TreeNode *root)
{
    if (root == nullptr)
    {
        BSTPair base;
        return base;
    }

    BSTPair lp = isBalanced_(root->left);
    BSTPair rp = isBalanced_(root->right);

    BSTPair myPair;
    myPair.isBal = lp.isBal && rp.isBal;
    if (myPair.isBal && abs(lp.h - rp.h) < 2)
    {
        myPair.h = max(lp.h, rp.h) + 1;
    }
    else
    {
        myPair.isBal = false;
    }

    return myPair;
}

bool isBalanced(TreeNode *root)
{
    return isBalanced_(root).isBal;
}

BSTPair largestBST_(TreeNode *root)
{
    if (root == nullptr)
    {
        BSTPair base;
        return base;
    }

    BSTPair lp = largestBST_(root->left);
    BSTPair rp = largestBST_(root->right);

    BSTPair myPair;
    BSTPair.isBST = false;
    if (lp.isBST && rp.isBST && lp.max < root->val && root->val < rp.min)
    {
        myPair.isBST = true;
        myPair.min = min(lp.min, root->val);
        myPair.max = max(rp.max, root->val);
        myPair.size = lp.size + rp.size + 1;
        myPair.largestRoot = root;
    }
    else
    {
        if (lp.size > rp.size)
        {
            myPair.size = lp.size;
            myPair.largestRoot = lp.largestRoot;
        }
        else
        {
            myPair.size = rp.size;
            myPair.largestRoot = rp.largestRoot;
        }
    }

    return myPair;
}

TreeNode *largestBST(TreeNode *root)
{
    return largestBST_(root).largestRoot;
}
